File: backend/app/services/segment_tree.py
"""
Segment Tree utilities
"""
from typing import List, Callable


class SegmentTree:
    """Segment tree implementation"""
    def __init__(self, arr: List[int], combine: Callable[[int, int], int]):
        self._n = len(arr)
        self._combine = combine
        self._tree = [0] * (4 * self._n)
        self._build(arr, 1, 0, self._n - 1)

    def _build(self, arr: List[int], node: int, start: int, end: int) -> None:
        """Build segment tree"""
        if start == end:
            self._tree[node] = arr[start]
        else:
            mid = (start + end) // 2
            self._build(arr, 2 * node, start, mid)
            self._build(arr, 2 * node + 1, mid + 1, end)
            self._tree[node] = self._combine(self._tree[2 * node], self._tree[2 * node + 1])

    def update(self, index: int, value: int) -> None:
        """Update value at index"""
        self._update_helper(1, 0, self._n - 1, index, value)

    def _update_helper(self, node: int, start: int, end: int, index: int, value: int) -> None:
        """Update helper"""
        if start == end:
            self._tree[node] = value
        else:
            mid = (start + end) // 2
            if index <= mid:
                self._update_helper(2 * node, start, mid, index, value)
            else:
                self._update_helper(2 * node + 1, mid + 1, end, index, value)
            self._tree[node] = self._combine(self._tree[2 * node], self._tree[2 * node + 1])

    def query(self, left: int, right: int) -> int:
        """Query range [left, right]"""
        return self._query_helper(1, 0, self._n - 1, left, right)

    def _query_helper(self, node: int, start: int, end: int, left: int, right: int) -> int:
        """Query helper"""
        if right < start or end < left:
            return 0
        if left <= start and end <= right:
            return self._tree[node]
        mid = (start + end) // 2
        left_result = self._query_helper(2 * node, start, mid, left, right)
        right_result = self._query_helper(2 * node + 1, mid + 1, end, left, right)
        return self._combine(left_result, right_result)

File: src/lib/segment-tree.ts
/**
 * Segment Tree utilities
 */

export class SegmentTree {
  private tree: number[]
  private n: number
  private combine: (a: number, b: number) => number

  constructor(arr: number[], combine: (a: number, b: number) => number) {
    this.n = arr.length
    this.combine = combine
    this.tree = new Array(4 * this.n).fill(0)
    this.build(arr, 1, 0, this.n - 1)
  }

  private build(arr: number[], node: number, start: number, end: number): void {
    if (start === end) {
      this.tree[node] = arr[start]
    } else {
      const mid = Math.floor((start + end) / 2)
      this.build(arr, 2 * node, start, mid)
      this.build(arr, 2 * node + 1, mid + 1, end)
      this.tree[node] = this.combine(this.tree[2 * node], this.tree[2 * node + 1])
    }
  }

  update(index: number, value: number): void {
    this.updateHelper(1, 0, this.n - 1, index, value)
  }

  private updateHelper(node: number, start: number, end: number, index: number, value: number): void {
    if (start === end) {
      this.tree[node] = value
    } else {
      const mid = Math.floor((start + end) / 2)
      if (index <= mid) {
        this.updateHelper(2 * node, start, mid, index, value)
      } else {
        this.updateHelper(2 * node + 1, mid + 1, end, index, value)
      }
      this.tree[node] = this.combine(this.tree[2 * node], this.tree[2 * node + 1])
    }
  }

  query(left: number, right: number): number {
    return this.queryHelper(1, 0, this.n - 1, left, right)
  }

  private queryHelper(node: number, start: number, end: number, left: number, right: number): number {
    if (right < start || end < left) {
      return 0
    }
    if (left <= start && end <= right) {
      return this.tree[node]
    }
    const mid = Math.floor((start + end) / 2)
    const leftResult = this.queryHelper(2 * node, start, mid, left, right)
    const rightResult = this.queryHelper(2 * node + 1, mid + 1, end, left, right)
    return this.combine(leftResult, rightResult)
  }
}
